index.coffee

```diff
@@ -7,27 +7,28 @@ basicAuthParser = require 'basic-auth-parser'
 MiddlewareHandler = require 'middleware-handler'
 MiddlewareHandler.prototype = Promise.promisifyAll(MiddlewareHandler.prototype)
 
-# Simplify testing
-exports.connect = net.connect
-
 # Connect an http socket to another tcp server.
 # Based on tunneling proxy code from https://nodejs.org/api/http.html
-connectSocket = (cltSocket, hostname, port, head) ->
-	srvSocket = exports.connect port, hostname, ->
-		cltSocket.write 'HTTP/1.0 200 Connection Established\r\n\
-				Proxy-agent: Resin-VPN\r\n\
-				\r\n'
+connectSocket = ({ cltSocket, hostname, port, head, connect, req }) ->
+	connect(port, hostname, cltSocket, req)
+	.then (srvSocket) ->
+		cltSocket.write('HTTP/1.0 200 Connection Established\r\nProxy-agent: Resin-VPN\r\n\r\n')
 		srvSocket.write(head)
 		srvSocket.pipe(cltSocket, end: false)
 		cltSocket.pipe(srvSocket, end: false)
-	Promise.fromNode (cb) ->
-		cltSocket.on('error', cb)
-		srvSocket.on('error', cb)
-		cltSocket.on('end', cb)
-		srvSocket.on('end', cb)
-	.finally (e) ->
-		srvSocket.destroy()
+
+		Promise.fromNode (cb) ->
+			cltSocket.on('error', cb)
+			srvSocket.on('error', cb)
+			cltSocket.on('end', cb)
+			srvSocket.on('end', cb)
+		.finally ->
+			srvSocket.destroy()
+			cltSocket.destroy()
+	.catch (e) ->
+		cltSocket.end('HTTP/1.0 500 Internal Server Error\r\n')
 		cltSocket.destroy()
+		throw e
```

Comment:
You can use
```coffee
.tapCatch ->
	cltSocket.end('HTTP/1.0 500 Internal Server Error\r\n')
	cltSocket.destroy()
```
instead now that we're using bluebird 3 and avoid having to manually rethrow